test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java

```diff
@@ -0,0 +1,241 @@
+package jenkins.model;
+
+import com.gargoylesoftware.htmlunit.html.HtmlForm;
+import hudson.Functions;
+import hudson.Util;
+import hudson.maven.MavenModuleSet;
+import hudson.maven.MavenModuleSetBuild;
+import hudson.model.FreeStyleBuild;
+import hudson.model.FreeStyleProject;
+import org.junit.After;
+import org.junit.Assume;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.jvnet.hudson.test.ExtractResourceSCM;
+import org.jvnet.hudson.test.Issue;
+import org.jvnet.hudson.test.LoggerRule;
+import org.jvnet.hudson.test.MockFolder;
+import org.jvnet.hudson.test.RestartableJenkinsRule;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.stream.Stream;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Since JENKINS-50164, Jenkins#workspacesDir and Jenkins#buildsDir had their associated UI deleted.
+ * So instead of configuring through the UI, we now have to use sysprops for this.
+ * <p>
+ * So this test class uses a {@link RestartableJenkinsRule} to check the behaviour of this sysprop being
+ * present or not between two restarts.
+ *
+ * @since TODO
+ */
+public class JenkinsBuildsAndWorkspacesDirectoriesTest {
```

Comment:
Seems to be missing a test for the following:

* Customer build/workspace are set in existing Jenkins config (from version without this change)
* Jenkins is upgrades
* Jenkins config is saved
* configuration should not be reset - existing custom values should be retained until they set the system property

Because otherwise people are going to not do that step and mess things up.

And this implies there is no way to go back to default - once you have set the system property if you want to revert to the defaults you need to know what they are and set the system property to it (removing the system property should not change things)...